### ISSUE TITLE
Android Pie 9 64-bit screencap timeout

### ISSUE THREAD

**Issue by Constrat** [MaaAssistantArknights/MaaAssistantArknights#6064]
### Before raising the issue...

- [X] I understand that Issues are for feedback and problem solving, not trolling the comments section, and will provide as much information as possible to help solve the problem.
- [X] I filled in a short, clear title so that developers could quickly identify the general problem when going through the issue list. Instead of "Some suggestions", "Stuck", etc.
- [X] I have checked the latest beta update and there is no mention of the bug being fixed.

### Description

For the last week I've tried digging more deeply around the screencap RawByNc timeouting in a specific environment: Android 9 (Pie) 64-bit.

Unfortunately I couldn't find anything so I just tried and tested some stuff:

### Case 1
```
[2023-08-23 21:05:21.603][INF][Px45e4][Tx9ac] Try to find the fastest way to screencap
[2023-08-23 21:05:41.604][INF][Px45e4][Tx9ac] Call ` C:\Users\David\Downloads\MAA\adb\adb.exe -s 127.0.0.1:5555 exec-out "screencap | nc -w 3 10.0.2.2 52882" ` ret 259 , cost 20001 ms , stdout size: 0 , socket size: 8294416
[2023-08-23 21:05:41.604][WRN][Px45e4][Tx9ac] data is empty!
[2023-08-23 21:05:41.604][INF][Px45e4][Tx9ac] RawByNc is not supported
[2023-08-23 21:05:41.837][INF][Px45e4][Tx9ac] Call ` C:\Users\David\Downloads\MAA\adb\adb.exe -s 127.0.0.1:5555 exec-out "screencap | gzip -1" ` ret 0 , cost 233 ms , stdout size: 477903 , socket size: 0
[2023-08-23 21:05:41.859][INF][Px45e4][Tx9ac] screencap_end_of_line is LF
[2023-08-23 21:05:41.859][INF][Px45e4][Tx9ac] RawWithGzip cost 254 ms
[2023-08-23 21:05:42.228][INF][Px45e4][Tx9ac] Call ` C:\Users\David\Downloads\MAA\adb\adb.exe -s 127.0.0.1:5555 exec-out screencap -p ` ret 0 , cost 369 ms , stdout size: 333390 , socket size: 0
[2023-08-23 21:05:42.253][INF][Px45e4][Tx9ac] screencap_end_of_line is LF
[2023-08-23 21:05:42.253][INF][Px45e4][Tx9ac] Encode cost 395 ms
[2023-08-23 21:05:42.253][INF][Px45e4][Tx9ac] The fastest way is RawWithGzip , cost: 254 ms
```
Here it says (correctly) that RawByNc is not supported and data is empty (this costs 20 seconds! A lot of time)

### Case 2
```
[2023-08-23 21:07:56.762][INF][Px4c10][Tx4bc0] Try to find the fastest way to screencap
[2023-08-23 21:08:16.763][WRN][Px4c10][Tx4bc0] Wait handles: [process_info.hProcess, pipeov.hEvent, sockov.hEvent] timeout.
[2023-08-23 21:08:16.763][INF][Px4c10][Tx4bc0] Call ` C:\Users\David\Downloads\MAA\adb\adb.exe -s 127.0.0.1:5555 exec-out "screencap | nc -w 3 10.0.2.2 52995" ` ret 0 , cost 20000 ms , stdout size: 0 , socket size: 8294416
[2023-08-23 21:08:16.767][INF][Px4c10][Tx4bc0] screencap_end_of_line is LF
[2023-08-23 21:08:16.768][INF][Px4c10][Tx4bc0] RawByNc cost 20005 ms
[2023-08-23 21:08:16.968][INF][Px4c10][Tx4bc0] Call ` C:\Users\David\Downloads\MAA\adb\adb.exe -s 127.0.0.1:5555 exec-out "screencap | gzip -1" ` ret 0 , cost 200 ms , stdout size: 473313 , socket size: 0
[2023-08-23 21:08:16.979][INF][Px4c10][Tx4bc0] screencap_end_of_line is LF
[2023-08-23 21:08:16.979][INF][Px4c10][Tx4bc0] RawWithGzip cost 211 ms
[2023-08-23 21:08:17.269][INF][Px4c10][Tx4bc0] Call ` C:\Users\David\Downloads\MAA\adb\adb.exe -s 127.0.0.1:5555 exec-out screencap -p ` ret 0 , cost 290 ms , stdout size: 332569 , socket size: 0
[2023-08-23 21:08:17.295][INF][Px4c10][Tx4bc0] screencap_end_of_line is LF
[2023-08-23 21:08:17.295][INF][Px4c10][Tx4bc0] Encode cost 315 ms
[2023-08-23 21:08:17.295][INF][Px4c10][Tx4bc0] The fastest way is RawWithGzip , cost: 211 ms
```
Here, we have a bit more intel on what really happens:
`Wait handles: [process_info.hProcess, pipeov.hEvent, sockov.hEvent] timeout`

I've looked around the Controller in MaaCore but I couldn't extract any valuable information from that.

With Bluestacks 5, I've tried two other instances with different versions: Nougat32 (Android 7.1.2) and Nougat64 (Android 7.1.1)
With these two versions everything worked correctly and RawByNc gave a decent cost of 200~300ms which is to be expected.

Unfortunately my C++ knowledge is very limited, but looking at the data I've collected is it possible to implement a version checker here https://github.com/MaaAssistantArknights/MaaAssistantArknights/blob/0acb6c7cf5d02870f11e5f17053c775ebed5033d/src/MaaCore/Controller/AdbController.cpp#L380-L381
`adb shell getprop ro.build.version.release` returns the Android Version number. I found this issue only on Android 9.

ref: #5184, #5426

**Comment by horror-proton** [comment]
#3854 also had this problem (it works fine in the beginning though).
I think it's logged from
https://github.com/MaaAssistantArknights/MaaAssistantArknights/blob/4b6b80d884bd59c60a9ce20f3ace8e852c54cf27/src/MaaCore/Controller/Platform/Win32IO.cpp
There is too much more win32 api than I can understand, I guess it means the command did not return for too long (due to some network issue?) @aa889788

**Comment by Constrat** [comment]
I've checked #3854 . Seems like it used RawWithNc even tho it had issues, so everytime it screen capped it had `Wait handles: [process_info.hProcess, pipeov.hEvent, sockov.hEvent] timeout.`.

**Comment by Constrat** [comment]
> due to some network issue

Network issues related to specific android versions. As I'm using Bluestacks I haven't tried with other emulators. Would be great to test if other Android 9 instances of other emulators have this problem.

**Comment by Rbqwow** [comment]
IDK if RawByNc works fine or not here, but Bluestack 5 P64 on my device always works nice with MAA. Maybe a certain computer only likes a certain emulator. :P

**Comment by Constrat** [comment]
Does it not give you 20 seconds of delay every time you click link start?

**Comment by Rbqwow** [comment]
Never. Here's the log if anything can help.
[asst.log](https://github.com/MaaAssistantArknights/MaaAssistantArknights/files/12868111/asst.log)

**Comment by Constrat** [comment]
@Rbqwow Sorry for the late response:
```ruby
[2023-10-11 12:08:20.803][INF][Px6990][Tx6ed0] Try to find the fastest way to screencap
[2023-10-11 12:08:40.803][WRN][Px6990][Tx6ed0] Wait handles: [process_info.hProcess, pipeov.hEvent, sockov.hEvent] timeout.
[2023-10-11 12:08:40.803][INF][Px6990][Tx6ed0] Call ` adb -s 127.0.0.1:5555 exec-out "screencap | nc -w 3 10.0.2.2 4698" ` ret 0 , cost 20000 ms , stdout size: 0 , socket size: 8294416
[2023-10-11 12:08:40.814][INF][Px6990][Tx6ed0] screencap_end_of_line is LF
[2023-10-11 12:08:40.815][INF][Px6990][Tx6ed0] RawByNc cost 20011 ms
```
As you can see it takes 20seconds just to "crash" which is not supposed to happen.
Pie64 seems to have the same problem as Bluestacks HYPER-V

**Comment by Constrat** [comment]
After A LOT of time, I've digged a bit more in the BlueStacks website:
Pie64 is compatible with Hyper-V, this means that it has the SAME issues as HYPER-V.

@horror-proton  @MistEO is there a way to detect a specific emulator / android version / x64 or x86? 
Adding another condition here
https://github.com/MaaAssistantArknights/MaaAssistantArknights/blob/74055b2f4bf0eb2924e45df8f916621f6846815c/src/MaaCore/Controller/AdbController.cpp#L380-L381

**Comment by Constrat** [comment]
Well, after many months it seems like I found the timeout.
fixed with c74368181852a1e7b263b26a2cb83c47f6446aa9
